File: January 2025/27.py
class LRUCache:
      
    # Constructor for initializing the cache capacity with the given value.
    def __init__(self, cap):
        self.cache = {}
        self.capacity = cap
        self.order = []
        
    # Function to return value corresponding to the key.
    def get(self, key):
        if key in self.cache:
            self.order.remove(key)
            self.order.append(key)
            return self.cache[key]
        else:
            return -1
        
    # Function for storing key-value pair.
    def put(self, key, value):
        if key in self.cache:
            self.order.remove(key)
        elif len(self.cache) >= self.capacity:
            oldest = self.order.pop(0)
            del self.cache[oldest]
        self.cache[key] = value
        self.order.append(key)

File: January 2025/27.cpp
#include <unordered_map>
#include <list>

class LRUCache {
  private:
    int capacity;
    std::unordered_map<int, std::pair<int, std::list<int>::iterator>> cache;
    std::list<int> order;
    
  public:
    // Constructor for initializing the cache capacity with the given value.
    LRUCache(int cap) : capacity(cap) {}

    // Function to return value corresponding to the key.
    int get(int key) {
        if (cache.find(key) != cache.end()) {
            order.erase(cache[key].second);
            order.push_front(key);
            cache[key].second = order.begin();
            return cache[key].first;
        } else {
            return -1;
        }
    }

    // Function for storing key-value pair.
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            order.erase(cache[key].second);
        } else if (cache.size() >= capacity) {
            int oldest = order.back();
            order.pop_back();
            cache.erase(oldest);
        }
        order.push_front(key);
        cache[key] = {value, order.begin()};
    }
};
